### PR TITLE
Avoid random lookup to determine if attribute has undefined value for…

### PR DESCRIPTION
**Pull request by baldersheim**
… a document.

Instead of comparing value against undefined value, compare enum reference.

@toregge PR

### CODE REVIEW

**Review by toregge — APPROVED** [review]

File: searchlib/src/vespa/searchlib/attribute/singleenumattribute.h [hunk 1]
```diff
@@ -110,6 +110,9 @@ class SingleValueEnumAttribute : public B, public SingleValueEnumAttributeBase {
     void onUpdateStat() override;
     void reclaim_memory(generation_t oldest_used_gen) override;
     void before_inc_generation(generation_t current_gen) override;
+    bool isUndefined(DocId doc) const override {
+        return acquire_enum_entry_ref(doc) == this->getEnumStore().get_default_value_ref().load_relaxed();
```
Comment (toregge) [inline searchlib/src/vespa/searchlib/attribute/singleenumattribute.h:114]: Note that during enum store compaction, isUndefined() might return false when it should return true, due to acquire_enum_entry_ref(doc) returning an EntryRef referencing the moved default value, while get_default_value_ref() returns an EntryRef referencing the original default value.

**Review by toregge — CHANGES_REQUESTED** [review]
I do not think we should accept IsUndefined() returning false when it should return true.

File: searchlib/src/vespa/searchlib/attribute/singleenumattribute.h (hunk 1 above)
Comment (baldersheim) [reply searchlib/src/vespa/searchlib/attribute/singleenumattribute.h:114]: hmm, that is probably something we do not want to allow.
Lets discuss options tomorrow.
Could the default_value_ref be fixed for the application lifetime ?

**Comment by baldersheim** [comment]
Will not pursue this one until further proof is found as to its cost.

**Pull request closed without merging**